### PR TITLE
[installer]: define the rules around the configuration

### LINKED ISSUE #15121: [installer|docs] Write a guide about the Installer config contract and how to handle breaking changes and new config version
In the Installer, the config acts as a contract between the installer and the user. We considered the option to introduce a new config version in case of breaking changes.

We should write an (internal) guide on what is considered as breaking change and how/when a new config version should be introduced.

The guide should at least answer the following questions:
- Who is the user of the installer (internal/external)? What expectations/requirements do they have for the stability the config?
- What is considered as breaking change? [A new mandatory config value](https://github.com/gitpod-io/gitpod/issues/15028)? A changed config value structure?
- What alternative strategies do we have to avoid breaking changes?
- When do we want to introduce a new config version?
- How do we introduce a new config version?
- How does the migration between config versions look like?

### PR DESCRIPTION
**Pull request by mrsimonemms**
## Description

Takes the discussion with @corneliusludmann (see [Notion](https://www.notion.so/gitpod/Installer-Configuration-Contract-c2a82878089e4ce5b2edccb6c92b7c4c)) and formalises as our requirements for the config surface

## Related Issue(s)

Fixes #15121

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@mrsimonemms: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Review by ArthurSens — COMMENTED** [review]
The documentation looks pretty good! And the best thing about this is that this document can be the source of truth for automated tests 🙂.


Do you think it is reasonable to automate testing of all the requirements written in this document? Or you think that keeping ourselves accountable is good enough?

File: install/installer/pkg/config/README.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+-   [There are different kinds of deployment](#there-are-different-kinds-of-deployment)
+
+### Requirements
+
+Requirements must be adhered to, without exception.
+
+#### It must maintain backwards compatibility
+
+The configuration surface maintains an `apiVersion` which defines the schema used. Backwards compatibility means that a `config.yaml` that is valid must **ALWAYS** be valid.
```
[the remark this answers is not in the document]
Comment (mrsimonemms) [reply install/installer/pkg/config/README.md:33]: A very good question. I'm certainly a fan of automation. I did briefly look into something like [Pact](https://pact.io/), although that's primarily about contracts delivered over HTTP - I'm sure something exists for Golang-only contract testing though (or could write some tests if not).

My general opinion is that this is a desirable goal and one I would like to take tangible steps towards achieving. That said, I don't necessarily have a clear way of achieving it in my mind so I'm very much interested in collaborating on.

File: install/installer/pkg/config/README.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+-   A parameter cannot have a `required` tag added to the validation if it has not previously been required. There is some nuance to this requirement - for example, in the registry config, the `external` parameter is only required if `inCluster` is `false`
+-   By default, the Installer maintains a strict conversion of the YAML into a Golang struct <sup>[[1](https://pkg.go.dev/sigs.k8s.io/yaml#UnmarshalStrict)]</sup>. Removing any parameter from the definition will cause the configuration to be unparsable
+
+Parameters can be renamed provided that:
+
+1. The old parameter with the current behaviour is retained
+2. The old parameter is marked as deprecated in the `CheckDeprecated` function
+3. If both the old and new parameters are configured, a conflict error is raised in the `CheckDeprecated` function
+4. It is recommended that a rewrite rule is provided from the old to the new parameter
```
[the remark this answers is not in the document]
Comment (mrsimonemms) [reply install/installer/pkg/config/README.md:45]: `3` and `4` already exist and I'm currently refactoring these so we can put lots more items in (see https://github.com/gitpod-io/gitpod/pull/15773)

File: install/installer/pkg/config/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArthurSens) [reply install/installer/pkg/config/README.md:33]: > That said, I don't necessarily have a clear way of achieving it in my mind so I'm very much interested in collaborating on.

Me neither 😅. Let's iterate over an RFC?

**Review by corneliusludmann — APPROVED** [review]
Love it! :heart: 

Can be merged like this from my side and socialized with the teams in Slack. :rocket:

**Pull request merged**